aten/src/ATen/gen_backend_select_register.py

```diff
@@ -0,0 +1,47 @@
+from code_template import CodeTemplate
+
+FUNCTION_REGISTRATION = CodeTemplate("""\
+.op(torch::RegisterOperators::options()
+  .schema("${schema_string}")
+  .impl_unboxedOnlyKernel<decltype(${function_name}), &${function_name}>(DispatchKey::BackendSelect)
+  .aliasAnalysis(AliasAnalysisKind::FROM_SCHEMA))
+""")
+
+FUNCTION_DEFINITION = CodeTemplate("""\
+Tensor ${function_name}(${method_formals}) {
+  DispatchKey key = options.computeDispatchKey();
+  static auto op = c10::Dispatcher::singleton().findSchemaOrThrow("aten::${name}", "${overload_name}");
+  return op.callUnboxedWithDispatchKey<${formals_types}>(key, ${type_method_actuals});
+}
+""")
+
+def register_backend_select_methods(declarations, top_env):
```

Comment:
I think it'd be more straightforward if this function simply took declarations and returned generated artifacts, rather than also taking a dictionary and adding the artifacts to it.

I know this pattern is easy to fall into because of the convention of calling `CodeTemplate.substitute()` with an environment, but heavy use of dynamic dictionaries instead of explicit params is one of the patterns that makes the codegen pipeline so hard to reason about - take a look at `gen_python_functions` now, it no longer uses env dictionaries at all, even in the calls to CodeTemplate.substitute(). I think it makes it quite a bit easier to track what's going (at least without intensive study :)

Reply:
Yeah, i totally agree. Good call. 
Updated